firestore/collgroupref.go

```diff
@@ -36,3 +45,59 @@ func newCollectionGroupRef(c *Client, dbPath, collectionID string) *CollectionGr
 		},
 	}
 }
+
+// PartitionQuery partitions a query by returning partition cursors that can be used to run
+// the query in parallel. The returned partition cursors are split points that
+// can be used by RunQuery as starting/end points for the query results.
+// Partition queries require explicit ordering by __name__ and
+// select all descendant collections.
+//
+// count is the desired maximum number of partition points. The number must be
+// strictly positive.
+func (c *CollectionGroupRef) PartitionQuery(ctx context.Context, count int) (Query, []Query, error) {
+	client, err := firestore.NewClient(ctx)
+	if err != nil {
+		return Query{}, nil, fmt.Errorf("firestore.NewClient: %v", err)
+	}
+	defer client.Close()
+	// Partition queries require explicit ordering by __name__ and
+	// select all descendant collections.
+	q := c.OrderBy(DocumentID, Asc)
+	q.allDescendants = true
+	strQuery, err := q.toProto()
+	if err != nil {
+		return Query{}, nil, err
+	}
+	req := &pb.PartitionQueryRequest{
+		Parent:         c.parentPath,
+		PartitionCount: int64(count),
+		QueryType: &pb.PartitionQueryRequest_StructuredQuery{
+			StructuredQuery: strQuery,
+		},
+	}
+	var partitions []Query
+	var docs []string
+	it := client.PartitionQuery(ctx, req)
+	for {
+		cursor, err := it.Next()
+		if err == iterator.Done {
+			break
+		}
+		if err != nil {
+			return q, nil, fmt.Errorf("PartitionQuery: %v", err)
+		}
+		doc := cursor.Values[0].GetReferenceValue()
+		docs = append(docs, doc)
+	}
+	for i, doc := range docs {
+		partition := q
+		if i == 0 {
+			partitions = append(partitions, partition.EndBefore(doc))
+		} else if i == len(docs)-1 {
+			partitions = append(partitions, partition.StartAt(doc))
+		} else {
+			partitions = append(partitions, partition.StartAt(doc).EndBefore(docs[i+1]))
+		}
+	}
```

Comment:
Not sure I understand. You're getting partitions with `PartitionQuery()`, aren't they are already prepared, with all of these End and Start points? I thought, you need just to iterate every partition into list and return them, or even return the iterator itself.

Reply:
Please, take a look closer to java example that I pushed here.
This method only returns cursors and `before`, not queries, and these cursors need to be passed to RunQuery. You can see that other languages (java, python, nodejs) doing the same.